### ISSUE TITLE
openmw SEGV with OSG 3.6.3

### ISSUE THREAD

**Issue by mimi1vx** [openscenegraph/OpenSceneGraph#655]
openmw issue - > https://gitlab.com/OpenMW/openmw/issues/4692

openmw engine crash with signal 11 if is used latest stable OSG.

crashlog : https://pastebin.com/vVmuHFdp

openmw community bisected commit which caused this problem to: https://github.com/openscenegraph/OpenSceneGraph/commit/eee5d5482e41290e7c210273f8c6bae489cedb76

**Comment by robertosfield** [comment]
I suspect the OSG commit just reveals an underlying bug elsewhere rather than directly causing the cash. Where the actual problem might be will only be resolvable by having a small test program that can reproduce the issue.

Could this be a threading issue?  I'm not familiar with OpenMW, is it being run multi-threaeded?  What happens when the the osgViewer part uses SingleThreaded threading model.  If it is running multi-threaded then it would be worth trying a SingleThreaded to see what happens.  If it resolves the crash then it's a hint that it could be a threading issue.

At this point there isn't anything I can do to help resolve the issue as the crash is too open ended.

**Comment by akortunov** [comment]
> If it resolves the crash then it's a hint that it could be a threading issue.

I tried to use the SingleThreaded model, and the game still crashes. The only difference is the stack trace (crash happens in the main thread).

**Comment by robertosfield** [comment]
Perhaps this is some form of dangling pointer issue then.  A stack trace
for each of the different types of crashes (single threaded vs
multi-threaded, might help show a pattern.

**Comment by akortunov** [comment]
Here the logs:
[openmw-crash.multi.log](https://github.com/openscenegraph/OpenSceneGraph/files/2552635/openmw-crash.multi.log)
[openmw-crash.single.log](https://github.com/openscenegraph/OpenSceneGraph/files/2552636/openmw-crash.single.log)

Note: the game can crash on different lines of the osg::GLBufferObject::compileBuffer() function.

Also I get some such warnings, which I do not have with older OSG versions:
```
Warning: detected OpenGL error 'invalid operation' after applying attribute Viewport
```
but it may be a completely separate issue.

**Comment by mp3butcher** [comment]
stack trace would be more relevant to debug

**Comment by akortunov** [comment]
OSG_VERTEX_BUFFER_HINT=VAO does not fix the crash.

Here are logs with OSG_NOTIFY_LEVEL=INFO and OSG_NOTIFY_LEVEL=DEBUG, from game start to moment of crash:
https://yadi.sk/d/DBXZzI-Gk4RZKg
https://yadi.sk/d/lVFiPgfYVsgxYA

I had to use an external storage since these logs are quite large.

**Comment by robertosfield** [comment]
On Tue, 6 Nov 2018 at 16:17, Andrei Kortunov <notifications@github.com>
wrote:

> OSG_VERTEX_BUFFER_HINT=VAO does not fix the crash.
>
> Here are logs with OSG_NOTIFY_LEVEL=INFO and OSG_NOTIFY_LEVEL=DEBUG, from
> game start to moment of crash:
> https://yadi.sk/d/DBXZzI-Gk4RZKg
> https://yadi.sk/d/lVFiPgfYVsgxYA
>
> I had to use an external storage since these logs are quite large.
>
At this point in time I don't have enough idea what might specifically
amiss to go trawling through massive console log output, there might be
some clues in there, but it's essentially a massive forest which we are
looking for a specific pin needle, which may well not be in these
particular forests.

For this type of problem that ideal way for me to be able to start looking
properly into the issue is to have a simple example case that illustrates
the bug then I can use this to home in on what the bug might be, fix it,
then use this example to confirm the bug has been fixed.  I'm not familiar
with OpenMW, and have a lot of other work to get on with so don't have the
scope for working out how to build OpenMW and then seeing if I can recreate
the problem.

If we can distil things down to small example that reproduces it then I'll
be happy to dive in and see how to address it and get a fix into the
OpenSceneGraph-3.6 branch.  I am also happy to fire up the OSG community to
start test with a view to pushing out a 3.6.4 stable release to wrap any
fixes resulting from this, and other fixes that have been already applied
since 3.6.3.

**Comment by mp3butcher** [comment]
echo $DISPLAY please?

**Comment by akortunov** [comment]
> what's says
echo $DISPLAY
since 18.04 it may had changed to 1:0...

I am on Gentoo, so DISPLAY is still ":0".

**Comment by mp3butcher** [comment]
I'm agree with Robert can you give us a clean gdb backtrace

**Comment by mp3butcher** [comment]
your seems to crash on _extensions->debugObjectLabel
could you try to add if(_extensions->glObjectLabel) before function call _extensions->debugObjectLabel
to see if func entry point as been found before its use

**Comment by robertosfield** [comment]
I have just do a quick diff of OpenSceneGraph-3.6.3 and 3.6 and there are fixes that might affect the GL warning your are seeing with OpenMW, but I don't think there is anything will affect the GL related object lifetime issues that look to be the cause of the crash.

What I can say though is the OpenSceneGraph-3.6 is the stepping stone towards a final 3.6.4 stable release that will (hopefully) have a fix for the crash you are seeing.  Any fixes applied will be done to the 3.6 branch, these would then need testing by the OpenMW team, then we can go for a 3.6.4 stable release.  

So as a first step could members of the OpenMW team/dev community test the OpenSceneGraph-3.6 branch.  Provide a stack trace of any crash and if possible provide a small example that reproduces the issue.  We can then have a bash at fixing the problem.

**Comment by robertosfield** [comment]
Is there any more progress with testing the 3.6 branch and characterizing this issue?  

I'd like to get resolve but don't have a concrete enough characterization/ability to reproduce the issue so can't do anything to help at this point.

**Comment by JDGBOLT** [comment]
I'm sorry that this is a bit late, I was the user who had originally done the bisecting down to that particular commit that fixed the crashing. I'm just a user, so I'm not too familiar with the internals of OSG and how OpenMW specifically uses it. Here is a backtrace using gdb with both OSG and OpenMW compiled with debugging symbols. Reverting the commit listed above I do believe 'fixes' it, but as you said, might be something that just showed an underlying problem. Sorry I can't be of more help at the moment:

Thread 18 "openmw" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffb9ffb700 (LWP 7247)]
0x00007ffff7d5723d in osg::GLBufferObject::compileBuffer (this=this@entry=0x7fffa824c370)
    at /usr/src/debug/openscenegraph/src/osg/BufferObject.cpp:223
223	/usr/src/debug/openscenegraph/src/osg/BufferObject.cpp: No such file or directory.
(gdb) bt
#0  0x00007ffff7d5723d in osg::GLBufferObject::compileBuffer (this=this@entry=0x7fffa824c370)
    at /usr/src/debug/openscenegraph/src/osg/BufferObject.cpp:223
#1  0x00007ffff7e23959 in osg::VertexArrayState::bindElementBufferObject (ebo=0x7fffa824c370, this=0x7fffa8009c20)
    at /usr/src/debug/openscenegraph/include/osg/VertexArrayState:94
#2  osg::DrawElementsUShort::draw (this=0x7fff641ca100, state=..., useVertexBufferObjects=<optimized out>)
    at /usr/src/debug/openscenegraph/src/osg/PrimitiveSet.cpp:264
#3  0x00007ffff7dadd82 in osg::Geometry::drawPrimitivesImplementation (this=this@entry=0x7fff6d4fea80, renderInfo=...)
    at /usr/src/debug/openscenegraph/src/osg/Geometry.cpp:996
#4  0x00007ffff7daec21 in osg::Geometry::drawImplementation (this=0x7fff6d4fea80, renderInfo=...)
    at /usr/src/debug/openscenegraph/src/osg/Geometry.cpp:897
#5  0x00007ffff7d9e5b0 in osg::Drawable::drawInner (renderInfo=..., this=0x7fff6d4fea80)
    at /usr/src/debug/openscenegraph/include/osg/Drawable:276
#6  osg::Drawable::draw (this=0x7fff6d4fea80, renderInfo=...) at /usr/src/debug/openscenegraph/src/osg/Drawable.cpp:707
#7  0x00007ffff7b03a48 in osgUtil::RenderLeaf::render (this=0x555563e51f60, renderInfo=..., previous=<optimized out>)
    at /usr/src/debug/openscenegraph/src/osgUtil/RenderLeaf.cpp:81
#8  0x00007ffff7afe615 in osgUtil::RenderBin::drawImplementation (this=0x555564ab99e0, renderInfo=..., 
    previous=@0x7fffb9ffa9c0: 0x5555659468a0) at /usr/src/debug/openscenegraph/src/osgUtil/RenderBin.cpp:487
#9  0x00007ffff7aff043 in osgUtil::RenderBin::draw (this=0x555564ab99e0, renderInfo=..., previous=<optimized out>)
    at /usr/src/debug/openscenegraph/src/osgUtil/RenderBin.cpp:430
#10 0x00007ffff7afe654 in osgUtil::RenderBin::drawImplementation (this=0x555564b9a2e0, renderInfo=..., 
    previous=@0x7fffb9ffa9c0: 0x5555659468a0) at /usr/src/debug/openscenegraph/src/osgUtil/RenderBin.cpp:517
#11 0x00007ffff7aff043 in osgUtil::RenderBin::draw (this=0x555564b9a2e0, renderInfo=..., previous=<optimized out>)
    at /usr/src/debug/openscenegraph/src/osgUtil/RenderBin.cpp:430
#12 0x00007ffff7afe654 in osgUtil::RenderBin::drawImplementation (this=this@entry=0x55555658b500, renderInfo=..., 
    previous=@0x7fffb9ffa9c0: 0x5555659468a0) at /usr/src/debug/openscenegraph/src/osgUtil/RenderBin.cpp:517
#13 0x00007ffff7b08eb6 in osgUtil::RenderStage::drawImplementation (this=0x55555658b500, renderInfo=..., 
    previous=@0x7fffb9ffa9c0: 0x5555659468a0) at /usr/src/debug/openscenegraph/src/osgUtil/RenderStage.cpp:1406
#14 0x00007ffff7aff043 in osgUtil::RenderBin::draw (this=this@entry=0x55555658b500, renderInfo=..., 
    previous=@0x7fffb9ffa9c0: 0x5555659468a0) at /usr/src/debug/openscenegraph/src/osgUtil/RenderBin.cpp:430
#15 0x00007ffff7b08237 in osgUtil::RenderStage::drawInner (this=0x55555658b500, renderInfo=..., 
    previous=@0x7fffb9ffa9c0: 0x5555659468a0, doCopyTexture=@0x7fffb9ffa8ef: false)
    at /usr/src/debug/openscenegraph/src/osgUtil/RenderStage.cpp:931
#16 0x00007ffff7b09db7 in osgUtil::RenderStage::draw (this=0x55555658b500, renderInfo=..., 
    previous=@0x7fffb9ffa9c0: 0x5555659468a0) at /usr/src/debug/openscenegraph/src/osgUtil/RenderStage.cpp:1242
#17 0x00007ffff7b11d01 in osgUtil::SceneView::draw (this=0x5555563601e0)
    at /usr/src/debug/openscenegraph/src/osgUtil/SceneView.cpp:1363
#18 0x00007ffff77b1521 in osgViewer::Renderer::draw (this=0x5555566c98e0)
    at /usr/src/debug/openscenegraph/src/osgViewer/Renderer.cpp:774
#19 0x00007ffff7dc7e29 in osg::GraphicsContext::runOperations (this=0x5555566308b0)
    at /usr/src/debug/openscenegraph/include/osg/Camera:577
#20 0x00007ffff7e1cf5e in osg::OperationThread::run (this=this@entry=0x5555568603d0)
    at /usr/src/debug/openscenegraph/include/osg/Observer:44
#21 0x00007ffff7dc9a1c in osg::GraphicsThread::run (this=0x5555568603d0)
    at /usr/src/debug/openscenegraph/src/osg/GraphicsThread.cpp:38
#22 0x00007ffff7c1ff39 in OpenThreads::ThreadPrivateActions::StartThread (data=0x5555568603e8)
    at /usr/src/debug/openscenegraph/src/OpenThreads/pthreads/PThread.cpp:213
#23 0x00007ffff56b6a9d in start_thread () from /usr/lib/libpthread.so.0
#24 0x00007ffff4b46b23 in clone () from /usr/lib/libc.so.6
(gdb)

**Comment by mimi1vx** [comment]
last live snapshot OSG still has issue present.

for openSUSE:Tumbleweed builds -> https://build.opensuse.org/project/show/home:mimi_vx:test

**Comment by mp3butcher** [comment]
@JDGBOLT 
We don't know how openmw use osg:which format is used to store data.....is openmw override osg job...
It would be great you could reproduce the bug in a sample code...
if usevbo i guess what could be easily done is to repopulate geometry vbos by calling setUseVertesBufferObject(false);setUseVertexBufferObject(true);
after geometry loading from file...but

**Comment by openscenegraph** [comment]
Any progress on creating a small same program that illustrates the problem?

It's not practical for us to go compiling large 3rd party applications and attempting to reproduce a bug that could be in that application or possibly in the OSG.

**Comment by robertosfield** [comment]
On Wed, 28 Nov 2018 at 15:28, Julien Valentin <notifications@github.com>
wrote:

> If it can help, I usually have this kind of crash when I fill Geometry
> attribs (or indexed primitiveset) with an empty Array...
>
Is there erroneous scene graph/GL usage that isn't caught till it hits the
driver?

Do you have any commits/PR's to any of your software that you resolved this
issue in?  This might be helpful is see what the specific problem case is.

Could something be done to detect this type of problem usage?

**Comment by mp3butcher** [comment]
mmmh after rereading this issue stack trace I doubt it's the same problem:
I usually crash on pr submission not in GLBuffer::compile ...
The last stack trace make me think of an issue with BO::getBufferData() returning null BD
which yield null pointer access at l.223 entry.dataSource->getModifiedCount() or l.229 entry.dataSource->asImage()
My guess is OpenMW may plays with the forbidden BD::setBufferIndex() function...

**Comment by orbea** [comment]
> My guess is OpenMW may plays with the forbidden BD::setBufferIndex() function...

I grepped the openmw codebase for `setBufferIndex` and found 0 results.

**Comment by mp3butcher** [comment]
Is there any mention to setBufferObject or setVertexBufferObject?

**Comment by orbea** [comment]
Yes, for `serBufferObject`.
```
$ grep -ri setVertexBufferObject
components/sceneutil/riggeometry.cpp:            vertexArray->setVertexBufferObject(vbo);
components/sceneutil/riggeometry.cpp:                normalArray->setVertexBufferObject(vbo);
components/sceneutil/riggeometry.cpp:            tangentArray->setVertexBufferObject(vbo);
components/sceneutil/morphgeometry.cpp:            vertexArray->setVertexBufferObject(vbo);
components/terrain/buffercache.cpp:        uvs->setVertexBufferObject(new osg::VertexBufferObject);
components/terrain/chunkmanager.cpp:    positions->setVertexBufferObject(vbo);
components/terrain/chunkmanager.cpp:    normals->setVertexBufferObject(vbo);
components/terrain/chunkmanager.cpp:    colors->setVertexBufferObject(vbo);
```
I'm not very familiar with the code base or qualified to fix it so I can't say if there is a problem here or not.

https://github.com/OpenMW/openmw

**Comment by mp3butcher** [comment]
I'm not familiar with OpenMW but is there any particular reason why RigGeometry is reimplemented?

**Comment by orbea** [comment]
I couldn't say, hopefully someone better in contact with the openmw developers can ask them to help investigate in this issue.

I just use openmw once in a while and noticed this issue when updating OpenSceneGraph.

**Comment by mp3butcher** [comment]
If i'm interpreting correctly the bt, the crash come from an Indexed primitiveset with a ebo problem..
You can:
1)try to restore osg auto setup
commenting out
components/terrain/buffercache.cpp l228
 buffer->setElementBufferObject(new osg::ElementBufferObject);
(and all unnecessary calls to setVertexBufferObject)
2)Force EBO setup for each of your DrawElementsXXX with a setElementBufferObject(new osg::ElementBufferObject);

**Comment by akortunov** [comment]
> Force EBO setup for each of your DrawElementsXXX with a setElementBufferObject(new osg::ElementBufferObject);

```
auto draw = new osg::DrawElementsUShort(osg::PrimitiveSet::TRIANGLES,
                                                                  data->triangles.size(),
                                                                  (unsigned short*)&data->triangles[0]);
draw->setElementBufferObject(new osg::ElementBufferObject);
geometry->addPrimitiveSet(draw);
```

Such approach just sometimes changes the stack trace - the game crashes in the osg::DrawElementsUShort::draw()

Commenting line in the buffercache.cpp did not fix anything.

Applying visitor before setSceneData() makes no sence since OpenMW changes scene during runtime. I tried to apply it every frame and it did not help - the game still crashes.

**Comment by akortunov** [comment]
Also there can be an issue with how OpenMW loads data from meshes - it creates nodes only once and caches them. If there is a need to place an another object with same mesh, it clones the cached node and attaches it to scene root.
I have no idea what will be with buffer objects after cloning.

**Comment by akortunov** [comment]
Just a wild guess what can happen here: different drawable instances can have shared buffers. 
When OSG removes one of such instances from scene, it releases these shared buffers and OpenMW crashes upon access to freed memory.
Prior to https://github.com/openscenegraph/OpenSceneGraph/commit/eee5d5482e41290e7c210273f8c6bae489cedb76 buffers were not released and there were not any crashes.

**Comment by mp3butcher** [comment]
clone(CopyOp) return a copy of the object but depending on copyop objects are copied by pointer (shallow_copy) or by value (deep_copy)
so if you clone a geometry with shallow copy all Geometry composites objects will NOT be duplicated
geom and geomcopy will share same arrays, stateset and so on (see osg::CopyOp to deep copy specific objects)

**Comment by akortunov** [comment]
> so if you clone a geometry with shallow copy all Geometry composites objects will NOT be duplicated
geom and geomcopy will share same arrays, stateset ,primsets, and so on (see osg::CopyOp to deep copy specific objects)

So in theory there should be a check if the array is used by another node before actually release it?

**Comment by mp3butcher** [comment]
as osganimation have changed a bit since 3.4, my other "voodoo" guess would be that your problem would come from you custom node (Rig and Morph AFAIK).
If you have a way to filter them try that..

**Comment by mp3butcher** [comment]
I've just built and run the game without any problem with my osg repo:
https://github.com/mp3butcher/OpenSceneGraph
So there's definitly something wrong with mainstream osg
There are just few diff in State, Drawable, geometry and vertexarraystate
you can try to substitute files in osg3.4 to find the bug

**Comment by mp3butcher** [comment]
solved

**Comment by JDGBOLT** [comment]
Well, I am sorry to inform you that this doesn't seem to have fixed the crash. Had a crash with the backtrace:
#0  0x00007ffff7ce9fed in osg::GLBufferObject::compileBuffer() () from /usr/lib/libosg.so.158
#1  0x00007ffff7e87949 in osg::VertexArrayState::setArray(osg::VertexArrayState::ArrayDispatch*, osg::State&, osg::Array const*) () from /usr/lib/libosg.so.158
#2  0x00007ffff7d485a8 in osg::Geometry::drawVertexArraysImplementation(osg::RenderInfo&) const () from /usr/lib/libosg.so.158
#3  0x00007ffff7d48ae5 in osg::Geometry::drawImplementation(osg::RenderInfo&) const () from /usr/lib/libosg.so.158
#4  0x00007ffff7a8cc10 in osgUtil::RenderLeaf::render(osg::RenderInfo&, osgUtil::RenderLeaf*) () from /usr/lib/libosgUtil.so.158
#5  0x00007ffff7a86bfd in osgUtil::RenderBin::drawImplementation(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /usr/lib/libosgUtil.so.158
#6  0x00007ffff7a87c03 in osgUtil::RenderBin::draw(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /usr/lib/libosgUtil.so.158
#7  0x00007ffff7a86c44 in osgUtil::RenderBin::drawImplementation(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /usr/lib/libosgUtil.so.158
#8  0x00007ffff7a87c03 in osgUtil::RenderBin::draw(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /usr/lib/libosgUtil.so.158
#9  0x00007ffff7a86c44 in osgUtil::RenderBin::drawImplementation(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /usr/lib/libosgUtil.so.158
#10 0x00007ffff7a92b98 in osgUtil::RenderStage::drawImplementation(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /usr/lib/libosgUtil.so.158
#11 0x00007ffff7a87c03 in osgUtil::RenderBin::draw(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /usr/lib/libosgUtil.so.158
#12 0x00007ffff7a91d47 in osgUtil::RenderStage::drawInner(osg::RenderInfo&, osgUtil::RenderLeaf*&, bool&) () from /usr/lib/libosgUtil.so.158
#13 0x00007ffff7a93b62 in osgUtil::RenderStage::draw(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /usr/lib/libosgUtil.so.158
#14 0x00007ffff7a9f737 in osgUtil::SceneView::draw() () from /usr/lib/libosgUtil.so.158
#15 0x00007ffff7728851 in osgViewer::Renderer::draw() () from /usr/lib/libosgViewer.so.158
#16 0x00007ffff7d63179 in osg::GraphicsContext::runOperations() () from /usr/lib/libosg.so.158
#17 0x00007ffff7def5ee in osg::OperationThread::run() () from /usr/lib/libosg.so.158
#18 0x00007ffff7d650ce in non-virtual thunk to osg::GraphicsThread::run() () from /usr/lib/libosg.so.158
#19 0x00007ffff7bb5fa9 in OpenThreads::ThreadPrivateActions::StartThread(void*) () from /usr/lib/libOpenThreads.so.21
#20 0x00007ffff5626a9d in start_thread () from /usr/lib/libpthread.so.0
#21 0x00007ffff4aadb23 in clone () from /usr/lib/libc.so.6

Though it's inconsistent where it occurs. Sometimes it was within the first minute of starting OpenMW, other times after several minutes or more. Thanks for the attempt at a fix though, not sure if this is a different cause now, as it does look like the backtrace is slightly different. I'll try compiling OSG and OpenMW with debugging symbols to try to get a better backtrace.

**Comment by mp3butcher** [comment]
then why does i achieve to play right now?

**Comment by JDGBOLT** [comment]
It's possible that it's being aggravated by mods, I think one where I have had success getting it to crash at least somewhat reliably is to install this mod: https://www.nexusmods.com/morrowind/mods/37795?tab=files , but it could also be an issue with the shadows patch I am using. Other than that I'm not really too sure, as it's inconsistent, and doesn't just happen right away, sometimes it can take a while, or is instant. Otherwise I would try perhaps an Apitrace, which is what helped last time there was a crash bug with OpenMW and OSG when running certain mods. Unfortunately that Dev that pointed that problem out, and coincidentally is the one who actually ported OpenMW to OSG himself has since left the project. Otherwise probably there could be more help in trying to find the root cause.

**Comment by sbrkopac** [comment]
The size of OpenMW doesn’t make it conducive to debugging easily. It’s important that any debug or issues are presented against vanilla OSG. If there are mods affecting it then it is much harder to narrow down.

I tried to clone OpenMW and it has so many deps that it is impossible to just get up and go with it. We will need to narrow down with an example or maybe Scrawl can help.

**Comment by JDGBOLT** [comment]
Unfortunately Scrawl has left the project and basically disappeared online, so that is not possible. And as he is the one who converted OpenMW from Ogre to OSG practically singlehandedly, that is a lot of knowledge that has disappeared as well. And I have tested against the OpenSceneGraph-3.6 branch, with the patch by mp3butcher applied, and no other changes. The crashes that I have experienced has had OSG be in all the function calls. The mods I refer to are just the one above which is added to OpenMW, which replaces some of the models and textures and seems to make the problem, at least for me, occur reliably. 

In case it is useful for anyone, here is a core dump and the binary used for it that, at least in kdevelop, allows me to see what variables are what in the call stack within the code:
https://drive.google.com/open?id=1YGA8uB0w4vHVMiRKO19Am07CHMRv9SyP

And this is an Apitrace of the game running then crashing, it's a bit nausea inducing and is 1440P resolution, but that was the way I could find that would reliably crash within a relatively short amount of time to not have it balloon to gigantic proportions:
https://drive.google.com/open?id=1Z8fFFbH5KaZjfnFQv9TxrUAB3QbGZTu1

Sorry I can't be of much more help, just not that exprienced with either engine or OpenGL in general to make heads or tails of what things do what.

**Comment by mp3butcher** [comment]
I don't understand what you're doing.
When I reproduced your bug,  the game didn't even launch the intro....With my patch I played MW all the night in debug whitout any problem.
please make clean install

**Comment by mp3butcher** [comment]
I apologize, I may have made a mistake building the first time with osg3.6. 
(The game didn't start the intro?!)
It make me think of a bug requiring patchin, but as I can't achieve to reproduce the crash, I think the patch is useless. 
As I achieve to build and run openmw with git osg 3.6.x, I think openmw users can do the same.
Feel free to contact me for build assistance but close this issue.
It's really messy and contain differents problems whose responsability is hard to determine (openmw or osg)

**Comment by openscenegraph** [comment]
It is best for production end user applications such as OpenMW to use stock stable releases of the OSG.  When we go through release testing cycles we spend 1 to 2 months calling for testing of the OSG itself, and how it's working with end user applications.  This is the time to make sure our up coming stable releases work fine for your application, if there are problems then this release cycle is a good time to get fixes checked in.

As my current main focus is the VSG project, my work on the OSG has moved to maintenance, so the nature next step for the OSG will be a OSG-3.6.4 patch release based on the OSG.3.6 banch.  The aim will be to retain backwards compatibility with 3.6.3.  It's not meant to be a release for adding features.  I'll currently don't have any plans to do major OSG new feature development for the next 12 months, so pushing lots of new features isn't something I'll be keen on.

I'll close this thread now as I don't think there is anything on the OSG for use to be aware of or action.  If it does turn out there is a bug in the 3.6.3 release or the 3.6 branch then please open an new issue.  The more info for guiding us to reproduce the problem the better.

**Comment by psi29a** [comment]
Does this mean that there will not another major/minor release of OSG?

Will there be a migration/upgrade path for applications wanting to leap to VSG from OSG? Will there perhaps be a compatibility layer that tries to map concepts of OSG on-top of VSG?

**Comment by openscenegraph** [comment]
There will be a 3.6.4 minor release of the OSG when there critical fixes that really need to be rolled out, or enough time passes that the minor fixes could do with rolling out.

A major release of the OSG will have to wait for the intensity of work on the VSG to back off.  I'm aiming for VSG-1.0 to be out in the second half of 2019.

The VSG is a different beast to the OSG so there won't be an easy means of wrapping OSG classes to VSG.  The OSG has the same DNA as the OSG so you'll see some similar themes, think cousin rather than sibling.  As part of the present work on the VSG I have written a osg2vsg utility library that will serve as a tool for converting OSG data into VSG equivalents.  Right now just osg::Image is mapped over.  I'll add node/geometry and some state in the next few months.  

The VSG related repos are found at:

   https://github.com/vsg-dev

I'll be firing off a post to the osg-users ML/forum this month to give everyone an update of where we are at, how it'll like dovetail with OSG project.

Migrating from OSG to VSG isn't something I am thinking too much about at this stage - my goal is to make the best scene graph possible based on C++17 and Vulkan, without worry about backwards compatibility.  

Once we know what the VSG-1.0 will look like it'll be easier to figure out how best to migrate across for projects that decide it's appropriate for them to do.  OpenGL/OSG willl both be around for as long as people want to use them so migrating isn't a necessity. people can migrate when they feel it's appropriate.  

For new projects it's a different choice, once VSG-1.0 is out it'll be a pretty compelling lib to use,: smaller, and significantly faster than the OSG and thoroughly modern.

**Comment by AnyOldName3** [comment]
I think I've got a rough outline of what's happening here.

* Since the bisected commit, the `Geometry` destructor calls `releaseGLObjects`. https://github.com/openscenegraph/OpenSceneGraph/blame/master/src/osg/Geometry.cpp#L86
* `Geometry`'s `reelaseGLObjects` calls `releaseGLObjects` on all its arrays. https://github.com/openscenegraph/OpenSceneGraph/blame/master/src/osg/Geometry.cpp#L733
* These arrays can be shared, so shouldn't be released if they are.

@akortunov already suggested this here https://github.com/openscenegraph/OpenSceneGraph/issues/655#issuecomment-442760440, but @mp3butcher mistakenly dismissed it as a possibility, thinking the release command wasn't forwarded to the arrays if the ref_ptr says they still have users. This isn't the case. The issue was nearly prevented by him here, though, where he realised anything based on `releaseGLObjects` would cause the shared array issue: https://github.com/openscenegraph/OpenSceneGraph/pull/587#issuecomment-412280987

Hopefully, this is enough to identify the issue.

**Comment by openscenegraph** [comment]
Could you create a small test example that reproduces the issue?  This will help reproduce the problem and then can be used as a unit test for any fix for the issue.

**Comment by mp3butcher** [comment]
```
osg::Geometry * geom= osg::createTexturedQuadGeometry(osg::Vec3f(-0.5,-0.5,-0.5),osg::Vec3f(1,0,0),osg::Vec3f(0,0,1));
   geom->setUseVertexBufferObjects(true);
  osg::Geometry *  geom2=           (osg::Geometry*  )geom->clone(osg::CopyOp::SHALLOW_COPY);
   osg::Group *gr=new osg::Group;
   gr->addChild(geom);
   gr->addChild(geom2);

   viewer.setSceneData(gr);
   viewer.realize();
    viewer.frame();//for the fun as ObjectVisitor already have do its stuff
    //assume contextid==0
         int contextid=   0;
    OSG_WARN<<"BufferObject uploaded suppose ok at frame 0:"<<
             ( geom->getVertexArray()->getVertexBufferObject()->getGLBufferObject(contextid) ?
                   "OK" : "BAD!!!!!!!!!" )<<std::endl;
    //remove  geom2 (no more ref to it)
    gr->removeChild(1);

    OSG_WARN<<"BufferObject uploaded suppose ok at  preframe 1:"<<
              ( geom->getVertexArray()->getVertexBufferObject()->getGLBufferObject(contextid) ?
                    "OK" : "BAD!!!!!!!!!" )<<std::endl;
    viewer.frame();
    OSG_WARN<<"BufferObject uploaded suppose ok at postframe 1:"<<
              ( geom->getVertexArray()->getVertexBufferObject()->getGLBufferObject(contextid) ?
                    "OK" : "BAD!!!!!!!!!" )<<std::endl;
```

**Comment by openscenegraph** [comment]
As it looks like we now have a succinct way to reproduce the problem, and @mp3butcher has create an separate Issue for it I'll close this PR as the signal to noise issue is not as good as a Issue focused on the specific underlying problem.  if you want to track progress than follow this new Issue. #674
